### PR TITLE
Bind outbound connection to address from configuration

### PR DESCRIPTION
**Pull request by yashkin**
For security reasons it would be great to support such feature as
binding outbound address, so nsqd could be firewalled on accepting
certain addresses, that reasonable in case when nsq-consumers or
producers are placed on machines with multiple interfaces.

### CODE REVIEW

**Comment by mreiferson** [comment]
I'm not necessarily opposed to this change but can't this be addressed outside of `go-nsq` by configuring routing tables?

**Comment by twmb** [comment]
This will try to bind all outgoing connections to the same address. Unless I'm mistaken, if you set LocalAddr, this will fail when you have more than one server you're trying to connect to, as this will try to open two connections with the same source addr.

**Comment by yashkin** [comment]
mreiferson, imagine machine with two NICs, when you start nsq listener or consumer, ip address of which NIC they will use? The answer is unpredictable, probably, it's possible to solve issue with iptables and masquerading of **all** outgoing ip addresses, but it's tricky, isn't it?

twmb, if you look into `consumer.go` you will find out, that it uses map of connections, where key is address of nsq daemon, so you can easily connect to multiple servers. Either, if you will use `LocalAddr` with port equal to *0*, like `config.Set("local_addr", "127.0.0.1:0")` os find a free one and bind your client to that address, so there is no problems with multiple connections (they all will use different endpoints).

File: consumer_test.go [hunk 1]
```diff
@@ -187,6 +192,11 @@ func consumerTest(t *testing.T, cb func(c *Config)) {
 		t.Fatal("should not be able to connect to the same NSQ twice")
 	}
 
+	conn := q.conns()[0]
+	if conn.conn.LocalAddr().String() != laddr {
+		t.Fatal("connection should be binded to the specified address:", conn.conn.LocalAddr())
```
Comment (twmb) [inline consumer_test.go]: should be `bound`

**Comment by twmb** [comment]
Cool, I hadn't read that yet. This looks fine to me other than `binded` should be `bound` in error messages. I'd change `ConnTimeout` to `DialTimeout` as the timeout is only used in dialing.

**Comment by yashkin** [comment]
Thanks, fixed.

**Comment by mreiferson** [comment]
> mreiferson, imagine machine with two NICs, when you start nsq listener or consumer, ip address of which NIC they will use? The answer is unpredictable, probably, it's possible to solve issue with iptables and masquerading of all outgoing ip addresses, but it's tricky, isn't it?

Still feels wrong to me.  I don't disagree that this patch addresses the issue, my concern is that this approach implies that _all_ of the client libraries need the same patch.  Whereas, if you handled this at the networking layer, it would apply to _anything_ connecting to nsqd.

Curious to hear @jehiah's thoughts...

**Comment by yashkin** [comment]
If you leave `LocalAddr` parameter equals to `nil` clients would behave the same as before. It's an option, not a requirement.

**Comment by jehiah** [comment]
certainly won't be needed in most cases but i'm fine exposing this. :thumbsup:

**Comment by mreiferson** [comment]
alright alright

File: config.go [hunk 2]
```diff
@@ -476,14 +485,16 @@ func valueTypeCoerce(v interface{}, typ reflect.Type) reflect.Value {
...
 		tval.SetFloat(val.Float())
+	default:
+		tval.Set(val)
 	}
 	return tval
 }
 
 func coerceString(v interface{}) (string, error) {
-	switch v.(type) {
+	switch v := v.(type) {
```
Comment (mreiferson) [inline config.go]: let's not shadow `v` here (and all the other instances below)

File: conn.go [hunk 3]
```diff
@@ -139,8 +139,13 @@ func (c *Conn) getLogger() (logger, LogLevel, string) {
 
 // Connect dials and bootstraps the nsqd connection
 // (including IDENTIFY) and returns the IdentifyResponse
-func (c *Conn) Connect() (*IdentifyResponse, error) {
-	conn, err := net.DialTimeout("tcp", c.addr, time.Second)
+func (c *Conn) Connect() (_ *IdentifyResponse, err error) {
```
Comment (mreiferson) [inline conn.go]: prefer not to have named returns

File: consumer_test.go [hunk 4]
```diff
@@ -223,3 +233,9 @@ func consumerTest(t *testing.T, cb func(c *Config)) {
 		t.Fatal("failed message not done")
 	}
 }
+
+func port() string {
```
Comment (mreiferson) [inline consumer_test.go]: should we listen on `:0`, read the port, close, then return that assumed open port?

File: config.go (hunk 2 above)
Comment (yashkin) [inline config.go]: it's golang idiom -reuse name of variables in a type switch statements
https://golang.org/doc/effective_go.html#type_switch

Comment (mreiferson) [inline config.go]: TIL - thanks!

**Comment by mreiferson** [comment]
@yashkin thanks, would you mind squashing down to one commit and we'll merge?

**Pull request merged**